rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/flow/TbRuleChainInputNode.java

```diff
@@ -48,17 +57,78 @@ public class TbRuleChainInputNode implements TbNode {
 
     private TbRuleChainInputNodeConfiguration config;
     private RuleChainId ruleChainId;
+    private RuleChainId currentRuleChainId;
+    private boolean forwardMsgToDefaultRuleChain;
 
     @Override
     public void init(TbContext ctx, TbNodeConfiguration configuration) throws TbNodeException {
         this.config = TbNodeUtils.convert(configuration, TbRuleChainInputNodeConfiguration.class);
-        this.ruleChainId = new RuleChainId(UUID.fromString(config.getRuleChainId()));
+        this.forwardMsgToDefaultRuleChain = config.isForwardMsgToDefaultRuleChain();
+        this.currentRuleChainId = ctx.getSelf().getRuleChainId();
+        if (forwardMsgToDefaultRuleChain) {
+            this.ruleChainId = getRootRuleChainId(ctx);
```

Comment:
What if we are in a root rule chain already?

Reply:
@ViacheslavKlimov If we are in the root chain already we got into this part of the code in onMsg() method:
```
           if (targetRuleChainId.equals(currentRuleChainId)) {
            ctx.tellFailure(msg, new RuntimeException("Forwarding messages to the current rule chain is not allowed!"));
            return;
        }
```